File: scripts/make.py
#!/usr/bin/env python

from optparse import OptionParser
from cStringIO import StringIO
from string import Template
import re, os, base64, mimetypes, codecs
try:
    import json
except:
    import simplejson as json

mimetypes.add_type('application/xml', '.tmx')
mimetypes.add_type('application/xml', '.tsx')
mimetypes.add_type('application/xml', '.plist')

TEXT_MIMETYPES = 'application/xml text/plain text/json application/json text/html'.split(' ')
CODE_MIMETYPES = 'text/javascript application/javascript application/x-javascript'.split(' ')

RESOURCE_TEMPLATE   = Template(u'__resources__["$resource"] = {meta: {mimetype: "$mimetype"}, data: $data};')

class Compiler(object):
    config = None
    output = 'cocos2d.js'
    main_module = 'main'
    extensions = ['js', 'tmx', 'tms', 'plist', 'gif', 'jpg', 'jpeg', 'png']
    header = u''
    footer = u''

    def __init__(self, config='make.js'):
        self.config = self.read_config(config)
        print "Ouputting to: ", self.output

    def read_config(self, config_file):
        print "Loading config:", config_file
        f = codecs.open(config_file, 'r', encoding='utf-8')
        config = self.read_json_file(config_file)
        self.output = config['output']
        self.main_module = config['main_module'] or 'main'
        self.extensions = config['extensions']

        return config

    def make(self):
        """
        Compile everything into a single script
        """
        code = self.header

        # Prepend app globals needed for resources
        code += '\n(function() {\n'
        code += 'var __main_module_name__ = %s;\n' % json.dumps(self.main_module)
        code += 'var __resources__ = [];\n'
        code += 'function __imageResource(data) { var img = new Image(); img.src = data; return img; };\n'
        for key, val in self.app_config_dict().items():
            code += 'var %s = %s;\n' % (key.upper(), json.dumps(val))


        # Add all the code
        for source, dest in self.config['paths'].items():
            code += self.make_path(source, dest)


        # Append module.js file -- this handles all the module loading
        module_js_path = os.path.join(os.path.dirname(os.path.realpath(__file__)), 'module.js')
        code += codecs.open(module_js_path, 'r', encoding='utf-8').read()

        code += '\n})();\n'

        code += self.footer

        return code


    def make_path(self, source_path, dest_path=None):
        """
        Compile everything at a path and return the code
        """
        if not dest_path:
            dest_path = source_path 
        print 'Building Path:', source_path, ' => ', dest_path
        
        code = ''
        files = self.scan_for_files(source_path)
        for src_file in files:
            if src_file in self.app_configs():
                # Skip config files because they're JSON not JavaScript
                continue

            dst_file = self.dst_for_src(src_file)
            mimetype = mimetypes.guess_type(src_file)[0]
            
            print 'Building File:', src_file, ' => ', dst_file
            code += '\n';
            code += RESOURCE_TEMPLATE.substitute({
                'mimetype': mimetype,
                'resource': dst_file,
                'data': self.make_resource(src_file),
            })

        return code

    def dst_for_src(self, path):
        for source, dest in self.config['paths'].items():
            if path.startswith(source):
                return re.sub('\/+', '/', re.sub(r'^' + source.replace('/', '\\/'), dest, path))

        return path

    def app_configs(self):
        """
        Returns an array of paths pointing to all the config.js files for the app
        """
        configs = [u'src/libs/cocos2d/config.js']
        for source, dest in self.config['paths'].items():
            c = os.path.join(source, 'config.js')
            if os.path.exists(c):
                configs.append(c)

        return configs

            

    def app_config_dict(self):
        """
        Reads all the app's config.js files and returns a dictionary of their values
        """
        vals = {}
        for config in self.app_configs():
            data = self.read_json_file(config)
            vals.update(data)


        return vals
        
    def read_json_file(self, path):
        j = codecs.open(path, 'r', encoding='utf-8').read()

        # Strip comments
        j = re.sub(r"\/\/.*", '', j)
        j = re.sub(re.compile(r"\/\*.*?\*\/", re.DOTALL), '', j)

        # Fix unquoted keys
        j = re.sub(r"{\s*(\w)", r'{"\1', j)
        j = re.sub(r",\s*(\w)", r',"\1', j)
        j = re.sub(r"(\w):", r'\1":', j)

        # Fix trailing comma
        j = re.sub(re.compile(r",\s+}", re.DOTALL), '}', j)

        return json.loads(j)


    def scan_for_files(self, path):
        """
        Scan for files to build and return them as an array
        """
        if os.path.isfile(path):
            return [path]

        found_files = []

        for root, dirs, files in os.walk(path):
            for f in files:
                if f[0] == '.':
                    # Skip hidden files
                    continue

                if self.extensions and os.path.splitext(f)[1][1:] not in self.extensions:
                    # Unwanted file extension
                    continue

                full_path = os.path.join(root, f)
                found_files.append(full_path)
                    
        return found_files

        

    def make_resource(self, filename):
        """
        Returns a resource string for adding to the __resources__ global
        """

        mimetype = mimetypes.guess_type(filename)[0]

        is_code = (mimetype in CODE_MIMETYPES)
        is_text = (mimetype in TEXT_MIMETYPES)
        is_image = (mimetype.split('/')[0] == 'image')

        if is_code:
            data = codecs.open(filename, 'r', encoding='utf-8').read()
            data = self.parse_supers(data)
            # Wrap code in function
            data = "function(exports, require, module, __filename, __dirname) {\n%s\n}" % data
        elif is_text:
            data = codecs.open(filename, 'r', encoding='utf-8').read()
            # Escape text by converting to json
            data = json.dumps(data)
        elif is_image:
            data = open(filename, 'rb')
            # Base64 encode image and create dataURL
            data = self.b64(data)
            data = '__imageResource("data:%s;base64,%s")' % (mimetype, data)
        else: # is_binary
            data = open(filename, 'rb')
            data = self.b64(data)
            data = '"%s"' % data

        return data
            
    def b64(self, data):
        """
        Base 64 encode binary data
        """
        output = StringIO()
        data = base64.encode(data, output)

        # Remote whitespace from string
        return re.sub('\s+', '', output.getvalue())

    def parse_supers(self, code):
        """
        Changes:
            @super('foo', 'bar');
            @super()
            @super

        Into:
            arguments.callee.base.call(this, 'foo', 'bar');
            arguments.callee.base.call(this);
            arguments.callee.base.apply(this, arguments);
        """

        def replace_super_apply(matches):
            return '%sarguments.callee.base.apply(this, arguments);' % (matches.group('indent'))

        def replace_super(matches):
            args = ''
            if matches.group('args'):
                args = ', %s' % matches.group('args')

            return '%sarguments.callee.base.call(this%s);' % (matches.group('indent'), args)

        super_re = re.compile('''(?P<indent>[ \t]*)@super\((?P<args>.*)\)\s*;''')
        code = super_re.sub(replace_super, code)

        super_re = re.compile('''(?P<indent>[ \t]*)@super\s*;''')
        code = super_re.sub(replace_super_apply, code)

        return code

def main():
    parser = OptionParser(usage="Usage: cocos make [options]")
    parser.add_option("-c", "--config", dest="config",
                      help="configuration file. Default is make.js", metavar="CONFIG")

    parser.add_option("-f", "--file", dest="output",
                      help="write code to FILE. Overrides config file", metavar="FILE")

    parser.add_option("-s", "--source", dest="input",
                      help="compile everything in SRC. Config is ignored if specified", metavar="SRC")

    (options, args) = parser.parse_args()

    compiler = Compiler(options.config or 'make.js')

    if options.input:
        code = compiler.make_path(options.input)
    else:
        code = compiler.make()

    output = options.output or compiler.output
    if output:
        print "Writing output to:", output
        o = codecs.open(output, 'w', encoding='utf-8')
        o.write(code)
        o.close()
    else:
        print code


if __name__ == "__main__":
    main()
    

File: lib/cocos/commands/make.js
/*globals require module exports process console __dirname*/
/*jslint undef: true, strict: true, white: true, newcap: true, indent: 4 */
"use strict";

var sys       = require('sys'),
    opts      = require('../opts'),
    fs        = require('fs'),
    path      = require('path'),
    Template  = require('../template').Template,
    mimetypes = require('../mimetypes');

var options = [
    {   short: 'f',
        long: 'file',
        description: 'File to write output to. Overrides config file',
        value: true },

    {   short: 'c',
        long: 'config',
        description: 'Configuration file. Default is make.json',
        value: true }

];

mimetypes.addType('application/xml', '.tmx');
mimetypes.addType('application/xml', '.tsx');
mimetypes.addType('application/xml', '.plist');

var RESOURCE_TEMPLATE = new Template('__resources__["$resource"] = {meta: {mimetype: "$mimetype"}, data: $data};');
var TEXT_MIMETYPES = 'application/xml text/plain text/json application/json text/html'.split(' ');
var CODE_MIMETYPES = 'text/javascript application/javascript application/x-javascript'.split(' ');

/**
 * Merge an number of objects together and return the result as a new object
 */
function merge() {
    var o = {};
    for (var i = 0, len = arguments.length; i < len; i++) {
        var obj = arguments[i];
        for (var x in obj) {
            if (obj.hasOwnProperty(x)) {
                o[x] = obj[x];
            }
        }
    }

    return o;
}

/**
 * @memberOf cocos.commands.make
 * @class Compile a cocos2d project into a single javascript file
 * @param {String} [configFile=make.json] The project's config filename
 */
function Compiler(configFile) {
    this.readConfig = function (configFile) {
        sys.puts('Loading config: ' + configFile);

        var config = this.readJSONFile(configFile);
        this.output = config.output;
        this.mainModule = config.mainModule || config.main_module;
        this.extensions = config.extensions;

        return config;
    };

    this.config = this.readConfig(configFile || 'make.json');
}
(function () /** @lends cocos.commands.make.Compiler# */{

    /**
     * Read a JSON file and clean up any comments, unquoted keys and trailing
     * commas before returning the object
     *
     * @param {String} filename Name of the JSON file to read
     * @returns {Object} The JSON object
     */
    this.readJSONFile = function (filename) {
        var j = fs.readFileSync(filename, 'utf8');

        // Strip comments
        j = j.replace(/\/\/.*/g, '');
        j = j.replace(/\/\*(.|[\n\r])*?\*\//mg, '');

        // Fix unquoted keys
        j = j.replace(/\{\s*(\w)/g, '{"$1');
        j = j.replace(/,(\s*)(\w)/g, ',$1"$2');
        j = j.replace(/(\w):/g, '$1":');

        // Fix trailing comma
        j = j.replace(/,\s+\}/mg, '}');

        return JSON.parse(j);
    };

    this.__defineGetter__('appConfigFiles', function () {
        if (this.appConfigFiles_) {
            return this.appConfigFiles_;
        }

        var configs = ['src/libs/cocos2d/config.json', 'cocos2d/src/libs/cocos2d/config.json'];

        for (var source in this.config.paths) {
            if (this.config.paths.hasOwnProperty(source)) {
                var dest = this.config.paths[source];

                var c = path.join(source, 'config.json');
                if (path.existsSync(c)) {
                    configs.push(c);
                }
            }
        }

        this.appConfigFiles_ = configs;

        return this.appConfigFiles_;
    });

    /**
     * Reads all the app's config.js files and returns an of object their
     * values
     */
    this.__defineGetter__('appConfig', function () {
        if (this.appConfig_) {
            return this.appConfig_;
        }

        var vals = {}, data;
        for (var i = 0, len = this.appConfigFiles.length; i < len; i++) {
            var config = this.appConfigFiles[i];
            if (path.existsSync(config)) {
                data = this.readJSONFile(config);
                vals = merge(vals, data);
            }
        }

        this.appConfig_ = vals;
        return this.appConfig_;
    });

    /**
     * Compile everything into a single script
     */
    this.make = function () {
        var code = this.header || '';
        code += '\n(function() {\n';
        code += 'var __main_module_name__ = ' + JSON.stringify(this.mainModule) + ';\n';
        code += 'var __resources__ = {};\n';
        code += 'function __imageResource(data) { var img = new Image(); img.src = data; return img; };\n';

        // Add config options
        for (var key in this.appConfig) {
            if (this.appConfig.hasOwnProperty(key)) {
                code += 'var ' + key.toUpperCase() + ' = ' + JSON.stringify(this.appConfig[key]) + ';\n';
            }
        }

        // Add all the code/images/etc
        for (var source in this.config.paths) {
            if (this.config.paths.hasOwnProperty(source)) {
                var dest = this.config.paths[source];
                code += this.makePath(source, dest);
            }
        }

        var module_js = path.join(__dirname, 'module_js');
        code += fs.readFileSync(module_js, 'utf8');

        code += '\n})();\n';

        code += this.footer || '';

        return code;
    };

    /**
     * Compile everything at a path and return the code
     * 
     * @param {String} source Path to compile
     * @param {String} [dest=source] Output path
     * @returns {String} Compiled javascript source code
     */
    this.makePath = function (source, dest) {
        sys.puts('Building Path: ' + source + ' => ' + dest);

        var code = '';
        var files = this.scanForFiles(source);
        for (var i = 0, len = files.length; i < len; i++) {
            var sourceFile = files[i];
            if (!!~this.appConfigFiles.indexOf(sourceFile)) {
                continue;
            }

            var destFile = this.destForSource(sourceFile),
                mimetype = mimetypes.guessType(sourceFile);
                

            sys.puts('Building File: ' + sourceFile + ' => ' + destFile);
            code += '\n';
            code += RESOURCE_TEMPLATE.substitute({
                'mimetype': mimetype,
                'resource': destFile,
                'data': this.makeResource(sourceFile)
            });
        }


        return code;
    };

    this.destForSource = function (path) {
        for (var source in this.config.paths) {
            if (this.config.paths.hasOwnProperty(source)) {
                var dest = this.config.paths[source];

                // Source starts with config path
                if (path.indexOf(source) === 0) {
                    return path.replace(new RegExp(source), dest).replace(/\/+/, '/');
                }
            }
        }

        return source;
    };

    this.makeResource = function (filename) {
        var mimetype = mimetypes.guessType(filename);

        var isCode = (!!~CODE_MIMETYPES.indexOf(mimetype)),
            isText = (!!~TEXT_MIMETYPES.indexOf(mimetype)),
            isImage = (mimetype.split('/')[0] == 'image');

        var data;
        if (isCode) {
            data = fs.readFileSync(filename, 'utf8');
            data = "function(exports, require, module, __filename, __dirname) {\n" + data + "\n}";
        } else if (isText) {
            data = JSON.stringify(fs.readFileSync(filename, 'utf8'));
        } else if (isImage) {
            data = fs.readFileSync(filename).toString('base64');
            data = '__imageResource("data:' + mimetype + ';base64,' + data + '")';
        } else /* isBinary */ {
            data = JSON.stringify(fs.readFileSync(filename).toString('base64'));
        }

        return data;
    };

    this.guessMimeType = function (filename) {
        return 'image/png';
    };

    /**
     * Scan for files to build and return them as an array
     *
     * @param {String} source Path to scan
     * @returns {String[]} List of filenames
     */
    this.scanForFiles = function (source) {

        var foundFiles = [];

        // If given a file rather than a directory they just return that
        if (fs.statSync(source).isFile()) {
            return [source];
        }

        // Find all files in directory
        var files = fs.readdirSync(source);
        for (var i = 0, len = files.length; i < len; i++) {
            var file = files[i];
            // Skip hidden files
            if (file[0] == '.') {
                continue;
            }

            var fullPath = path.join(source, file);

            if (fs.statSync(fullPath).isFile()) {
                // If extension isn't in our list then skip file
                if (this.extensions && this.extensions.length && !~this.extensions.indexOf(path.extname(file).slice(1))) {
                    continue;
                }

                foundFiles.push(fullPath);
            } else {
                // Directory
                foundFiles = foundFiles.concat(this.scanForFiles(fullPath));
            }

        }
        
        return foundFiles;
    };
}).call(Compiler.prototype);

exports.Compiler = Compiler;

exports.description = 'Compile a cocos2d project into a single javascript file';
exports.run = function () {
    opts.parse(options, true);

    var config   = opts.get('config') || 'make.json',
        compiler = new Compiler(config),
        output   = opts.get('file')   || compiler.output;

    var code = compiler.make();

    if (output) {
        sys.puts("Writing output to: " + output);
        fs.writeFileSync(output, code, 'utf8');
    } else {
        sys.puts(code);
    }
};
